url1a745d67fc_sobercheg_InterviewTraining_tgz-pJ8-leetcode_oj_Solution2TestJ8/src/leetcode/oj/Solution.java

```diff
@@ -0,0 +1,3742 @@
+            ArrayList<Integer> permut = new ArrayList<Integer>();
+            for (int n : num) permut.add(n);
+            result.add(permut);
+            return result;
+        }
+
+        ArrayList<ArrayList<Integer>> result = new ArrayList<ArrayList<Integer>>();
+        Set<Integer> used = new HashSet<Integer>();
+        for (int i = level; i < num.length; i++) {
+            if (!keepUniqueIfSorted && used.contains(num[i])) continue;
+            used.add(num[i]);
+            exch(num, level, i);
+            result.addAll(permuteBacktracking(num, level + 1, keepUniqueIfSorted));
+            exch(num, level, i);
+        }
+
+        return result;
+
+    }
+
+    private void exch(int[] arr, int i, int j) {
+        int tmp = arr[i];
+        arr[i] = arr[j];
+        arr[j] = tmp;
+    }
+
+    /**
+     * <a href="http://en.wikipedia.org/wiki/Heap%27s_algorithm">Heap's algorithm</a>
+     * Suppose we have a sequence of different characters with a length of N.
+     * Heap found that we can simply interchange the positions of two elements to get a new permutation output.
+     * Let us describe it in a recursive way. If we have got (N - 1)! permutation outputs, fixing the last element.
+     * Then if N is odd, we can switch the first element and the last one, while N is even we can switch the ith (i
+     * is the step number of the cycle, and now it is 1) element and the last one, then we will continue outputting
+     * the (N - 1)! permutation outputs and switching step for another N − 1 times(N times int total).
+     * The following pseudocode outputs permutations of a data array of length N.
+     * <p/>
+     * procedure generate(N : integer, data : array of any):
+     * if N = 1 then
+     * output(data)
+     * else
+     * for c := 1; c <= N; c += 1 do
+     * generate(N - 1, data)
+     * swap(data[if N is odd then 1 else c], data[N])
+     */
+    private ArrayList<ArrayList<Integer>> permuteHeap(int[] num, int level) {
+        if (level == num.length) {
+            ArrayList<ArrayList<Integer>> result = new ArrayList<ArrayList<Integer>>();
+            ArrayList<Integer> permut = new ArrayList<Integer>();
+            for (int n : num) permut.add(n);
+            result.add(permut);
+            return result;
+        }
+        ArrayList<ArrayList<Integer>> result = new ArrayList<ArrayList<Integer>>();
+        for (int i = 0; i <= level; i++) {
+            result.addAll(permuteHeap(num, level + 1));
+            exch(num, level, level % 2 == 0 ? i : 0);
+        }
+
+        return result;
+    }
+
+    /**
+     * <a href="http://oj.leetcode.com/problems/rotate-image/">Rotate Image</a>
+     * You are given an n x n 2D matrix representing an image.
+     * <p/>
+     * Rotate the image by 90 degrees (clockwise).
+     * <p/>
+     * Follow up:
+     * Could you do this in-place?
+     */
+    public void rotate(int[][] matrix) {
+        int n = matrix.length;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;matrix&#39; (index: 0 | type: int[][). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other methods validating the parameter: &#39;spiralOrder&#39;. The same parameter type is validated here for example: njr_failed/url1a745d67fc_sobercheg_InterviewTraining_tgz-pJ8-leetcode_oj_Solution2TestJ8/src/leetcode/oj/Solution.java:3687. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).